src/coreclr/src/vm/jithelpers.cpp

```diff
@@ -5133,6 +5133,33 @@ Thread * __stdcall JIT_InitPInvokeFrame(InlinedCallFrame *pFrame, PTR_VOID StubS
 EXTERN_C void JIT_PInvokeBegin(InlinedCallFrame* pFrame);
 EXTERN_C void JIT_PInvokeEnd(InlinedCallFrame* pFrame);
 
+// Forward declaration
+EXTERN_C void STDCALL ReversePInvokeBadTransition();
+
+EXTERN_C void JIT_ReversePInvokeEnter(FastReversePInvokeFrame* frame)
+{
+    _ASSERTE(frame != NULL);
+
+    Thread* thread = GetThreadNULLOk();
+    if (thread == NULL)
+        thread = CreateThreadBlockThrow();
```

Comment:
There is a probably a way to make this faster by moving all rare cases into a separate `JIT_ReversePInvokeEnterRare` method.

Reply:
Yes, you were right. The amount of generated code was substantially reduced by reorganizing the branches.